File: 125 Valid Palindrome.swift
//leetcode link:
//https://leetcode.com/problems/valid-palindrome/

class Solution {
    func isPalindrome(_ s: String) -> Bool {
        let s = s.lowercased()
        var left = s.startIndex
        var right = s.index(before: s.endIndex)
        
        while left < right {
            while left < right && !s[left].isLetter && !s[left].isNumber {
                left = s.index(after: left)
            }
            while left < right && !s[right].isLetter && !s[right].isNumber {
                right = s.index(before: right)
            }
            if s[left] != s[right] {
                return false
            }
            left = s.index(after: left)
            right = s.index(before: right)
        }
        return true
    }
}

let sol = Solution()
print(sol.isPalindrome("A man, a plan, a canal: Panama"))  // ➜ true

File: 125 Valid Palindrome.py
#leetcode link:
# https://leetcode.com/problems/valid-palindrome/

class Solution:
    def isPalindrome(self, s: str) -> bool:
        l,r = 0, len(s)-1
        s = s.lower()
        
        while l <= r:
            while not s[l].isalnum() and l < r: l += 1
            while not s[r].isalnum() and l < r: r -= 1
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True
    
Input = "A man, a plan, a canal: Panama"

s = Solution()
print(s.isPalindrome(Input))